rtc/KalmanFilter/EKFilter.h

```diff
@@ -142,61 +99,50 @@ class EKFilter {
-    x_a_priori = calcPredictedState(q, u, drift, dt);
-    P_a_priori = calcPredictedCovariance(F);
+    Eigen::Matrix<double, 7, 7> F;
+    calcF(F, q, u, drift);
+    Eigen::Matrix<double, 7, 1> x_tmp;
+    calcPredictedState(x_tmp, q, u, drift);
+    x_a_priori = x_tmp;
+    Eigen::Matrix<double, 7, 7> P_tmp;
+    calcPredictedCovariance(P_tmp, F, q);
+    P_a_priori = P_tmp;
   }
 
-  void correction(const Eigen::Vector3d& z,
-                  const Eigen::Vector3d& vel_ref,
-                  const Eigen::Vector3d& acc_ref,
-                  const Eigen::Vector3d& angular_rate_ref) {
-    Eigen::Matrix<double, 4, 1> q_a_priori = x_a_priori.block<4, 1>(0, 0).normalized();
+  void correction(const Eigen::Vector3d& z) {
+    Eigen::Matrix<double, 4, 1> q_a_priori = x_a_priori.block<4, 1>(0, 0);
     Eigen::Matrix<double, 3, 7> H;
-    Eigen::Matrix<double, 3, 3> S;
-    Eigen::Matrix<double, 7, 3> K;
-    Eigen::Vector3d y;
-    /* need to normalize q_a_priori ? */
-    y = calcMeasurementResidual(z, vel_ref, acc_ref, angular_rate_ref, q_a_priori);
-    H = calcH(q_a_priori);
-    S = H * P_a_priori * H.transpose() + R;
-    K = P_a_priori * H.transpose() * S.inverse();
+    Eigen::Vector3d y = calcMeasurementResidual(z, q_a_priori);
+    calcH(H, q_a_priori);
+    Eigen::Matrix<double, 3, 3> S = H * P_a_priori * H.transpose() + R;
+    Eigen::Matrix<double, 7, 3> K = P_a_priori * H.transpose() * S.inverse();
     Eigen::Matrix<double, 7, 1> x_tmp = x_a_priori + K * y;
-    x = x_tmp.normalized();
+    x.block<4, 1>(0, 0) = x_tmp.block<4, 1>(0, 0).normalized(); /* quaternion */
+    x.block<3, 1>(4, 0) = x_tmp.block<3, 1>(4, 0); /* bias */
     P = (Eigen::Matrix<double, 7, 7>::Identity() - K * H) * P_a_priori;
   }
 
-  void printAll() {
+  void printAll() const {
     std::cerr << "x" << std::endl << x << std::endl;
     std::cerr << "x_a_priori" << std::endl << x_a_priori << std::endl;
     std::cerr << "P" << std::endl << P << std::endl << std::endl;
     std::cerr << "P_a_priori" << std::endl << P_a_priori << std::endl << std::endl;
-    /*
-     * std::cerr << "Q" << std::endl << Q << std::endl << std::endl;
-     * std::cerr << "R" << std::endl << R << std::endl << std::endl;
-     */
   }
 
   void main_one (hrp::Vector3& rpy, hrp::Vector3& rpyRaw, const hrp::Vector3& acc, const hrp::Vector3& gyro)
   {
-      prediction(gyro);
-      correction(acc, Eigen::Vector3d(0, 0, 0), Eigen::Vector3d(0, 0, 0), Eigen::Vector3d(0, 0, 0));
-      /* ekf_filter.printAll(); */
-      Eigen::Matrix<double, 7, 1> x = getx();
-      Eigen::Quaternion<double> q = Eigen::Quaternion<double>(x[0], x[1], x[2], x[3]);
-      hrp::Vector3 eulerZYX = q.toRotationMatrix().eulerAngles(2,1,0);
-      rpy(2) = eulerZYX(0);
-      rpy(1) = eulerZYX(1);
-      rpy(0) = eulerZYX(2);
+    prediction(gyro);
+    correction(acc);
+    /* ekf_filter.printAll(); */
+    Eigen::Quaternion<double> q = Eigen::Quaternion<double>(x[0], x[1], x[2], x[3]);
```

Comment:
Can you rather use the quarternion constructor that takes a vector 4 as argument?